### PR TITLE
fix: Preserve table cell styles on Google Docs paste

### PR DESCRIPTION
**Pull request by Branc0**
### Summary
PR for the issue [Issue #2150](https://github.com/superdoc-dev/superdoc/issues/2150).

Table cell styling is lost when pasting from Google Docs. Cells with background colors are pasted as plain white; borders and padding are dropped.

Additionally, as a side effect of the main fixes, this PR solves two other issues:
- Google Docs Pasted tables don't persist row height: I had to fix to test vertical alignment.
- Duplicated borders on the intersections of cells. As we make the borders bigger, this gets more evident. 

### Steps for testing
- [ ] Create a table in Google Docs with colored cells (e.g., [this doc](https://docs.google.com/document/d/1t4ZZOPmQm8YMXFmPLlQi4zXJNS52YeEB19TefSIFlu0/edit?tab=t.0) ).
- [ ] Copy the table and paste it into SuperDoc.
- [ ] Table cell background colors, borders, padding, and vertical alignment from pasted content are preserved.

### CODE REVIEW

File: packages/super-editor/src/editors/v1/extensions/table-cell/helpers/parseCellMargins.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ */
+export const parseCellMargins = (element) => {
+  const { style } = element;
+
+  const [topValuePt] = parseSizeUnit(style?.paddingTop);
+  const [rightValuePt] = parseSizeUnit(style?.paddingRight);
+  const [bottomValuePt] = parseSizeUnit(style?.paddingBottom);
+  const [leftValuePt] = parseSizeUnit(style?.paddingLeft);
+
+  const top = halfPointToPixels(topValuePt);
```
Comment (Branc0) [inline packages/super-editor/src/editors/v1/extensions/table-cell/helpers/parseCellMargins.js]: I was searching for a helper function that converts PT to PX, and I noticed this has the same formula, so I used it despite the name being halfPoint. Let me know if it's okay...

File: packages/super-editor/src/editors/v1/extensions/table-cell/table-cell.js [hunk 2]
```diff
@@ -164,6 +172,13 @@ export const TableCell = Node.create({
       },
 
       verticalAlign: {
+        parseDOM: (element) => {
+          const value = element.style?.verticalAlign;
+          if (!value || typeof value !== 'string') return null;
+          const normalized = value.trim().toLowerCase();
+          if (normalized === 'middle') return 'center';
```
Comment (Branc0) [inline packages/super-editor/src/editors/v1/extensions/table-cell/table-cell.js]: In Google Docs, it's used 'middle'  to vertically center content in the cell

File: packages/super-editor/src/editors/v1/extensions/table-row/helpers/parseRowHeight.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+/**
+ * Parse row height from a <tr> element.
+ * Priority: row height/min-height/height attr, then tallest td/th height.
+ *
+ * @param {HTMLElement} element
+ * @returns {number | undefined}
+ */
+export const parseRowHeight = (element) => {
+  return parseExplicitRowHeight(element) ?? parseTallestCellHeight(element);
```
Comment (Branc0) [inline packages/super-editor/src/editors/v1/extensions/table-row/helpers/parseRowHeight.js:81]: The Idea here was to get the height first from the inline style of the row; if there's no explicit styling there, it switches the approach and picks from the highest cell instead.

File: packages/super-editor/src/editors/v1/extensions/shared/parseCellBorders.js [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  'none',
+  'hidden',
+  'dotted',
+  'dashed',
+  'solid',
+  'double',
+  'groove',
+  'ridge',
+  'inset',
+  'outset',
```
Comment (Branc0) [inline packages/super-editor/src/editors/v1/extensions/shared/parseCellBorders.js:34]: This may be overkill since Google Docs only supports 'solid', 'dashed', and 'dotted'. I introduced the other options in case HTML gets pasted.

File: packages/layout-engine/painters/dom/src/table/renderTableRow.ts [hunk 5]
```diff
@@ -74,6 +74,20 @@ const resolveRenderedCellBorders = ({
...
+      return {
+        top: resolveTableBorderValue(cellBorders.top, touchesTopBoundary ? tableBorders.top : tableBorders.insideH),
+        right: cellBounds.touchesRightEdge ? resolveTableBorderValue(cellBorders.right, tableBorders.right) : undefined,
+        bottom: touchesBottomBoundary ? resolveTableBorderValue(cellBorders.bottom, tableBorders.bottom) : undefined,
+        left: resolveTableBorderValue(
+          cellBorders.left,
+          cellBounds.touchesLeftEdge ? tableBorders.left : tableBorders.insideV,
+        ),
+      };
+    }
```
Comment (Branc0) [inline packages/layout-engine/painters/dom/src/table/renderTableRow.ts:91]: The idea here is to prevent this issue from happening: 

Image

If I understood correctly, beneath the hood, the table would still have the styles applied in the cells, so I'm not sure if this has the risk of being rendered incorrectly when exported, for example.

**Comment by Branc0** [comment]
Still WIP on this one:
- The P1 issue from codex: tried a couple of fixes, but they introduced the duplicated border back
- Integration test

File: packages/super-editor/src/editors/v1/extensions/shared/parseCellBorders.js [hunk 6]
```diff
@@ -0,0 +1,139 @@
+// @ts-check
```
Comment (caio-pizzol) [inline packages/super-editor/src/editors/v1/extensions/shared/parseCellBorders.js:1]: lots of parsing branches here but only one test. a few more covering `none`, per-side borders, `dashed`/`dotted`, and different colors would help — worth adding?

File: packages/layout-engine/painters/dom/src/table/renderTableRow.ts [hunk 7]
```diff
@@ -74,6 +74,20 @@ const resolveRenderedCellBorders = ({
   const touchesBottomBoundary = cellBounds.touchesBottomEdge || continuesOnNext;
 
   if (hasExplicitBorders) {
+    if (cellSpacingPx === 0) {
+      // Collapsed model: avoid double interior borders by using single-owner sides.
+      // Keep explicit top/left (or table fallbacks), and only render right/bottom on table edges.
+      return {
+        top: resolveTableBorderValue(cellBorders.top, touchesTopBoundary ? tableBorders.top : tableBorders.insideH),
+        right: cellBounds.touchesRightEdge ? resolveTableBorderValue(cellBorders.right, tableBorders.right) : undefined,
+        bottom: touchesBottomBoundary ? resolveTableBorderValue(cellBorders.bottom, tableBorders.bottom) : undefined,
```
Comment (caio-pizzol) [inline packages/layout-engine/painters/dom/src/table/renderTableRow.ts:85]: re: your concern about dropped borders — tested it and GDocs always puts the same border on both sides, so this works. could be an issue with asymmetric borders from other sources, worth a comment noting the assumption.

File: packages/super-editor/src/editors/v1/extensions/shared/parseCellMargins.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
+// @ts-check
```
Comment (caio-pizzol) [inline packages/super-editor/src/editors/v1/extensions/shared/parseCellMargins.js:1]: re: `halfPointToPixels` — same formula is now in three files. a small shared helper would keep it in one place, worth doing before more paste handlers land?

**Review by caio-pizzol — COMMENTED** [review]
@Branc0 tested with real Google Docs tables — backgrounds, borders (solid/dashed/dotted), merged cells, borderless cells, padding, column widths all working. no blockers.

test docs: [basic styles](https://docs.google.com/document/d/1LUv_C2WRAD76Pc7h4hlB6zUqwDSMu6XCsW3lH7gpIV8/edit) · [edge cases](https://docs.google.com/document/d/1XPZmDIX8pgbjOAZntO0nvNk8ERZZGLkwIdv8B0JzVMc/edit)

basic styles — GDocs vs SuperDoc:
image
image

edge cases — GDocs vs SuperDoc:
image
image


left a few inline comments.

File: packages/layout-engine/painters/dom/src/table/renderTableRow.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (caio-pizzol) [reply packages/layout-engine/painters/dom/src/table/renderTableRow.ts:85]: @Branc0 this is intentional — Google Docs always puts the same border on both sides, so nothing gets lost.

File: packages/super-editor/src/editors/v1/extensions/table-cell/helpers/parseCellMargins.js [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+  const right = halfPointToPixels(rightValuePt);
+  const bottom = halfPointToPixels(bottomValuePt);
+  const left = halfPointToPixels(leftValuePt);
```
[the remark this answers is not in the document]
Comment (caio-pizzol) [reply packages/super-editor/src/editors/v1/extensions/table-cell/helpers/parseCellMargins.js]: @Branc0 already fixed — now checks the unit before converting.

File: packages/super-editor/src/editors/v1/extensions/table-cell/helpers/parseCellMargins.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+  return {
+    top: top ?? 0,
+    right: right ?? 0,
+    bottom: bottom ?? 0,
+    left: left ?? 0,
```
[the remark this answers is not in the document]
Comment (caio-pizzol) [reply packages/super-editor/src/editors/v1/extensions/table-cell/helpers/parseCellMargins.js]: @Branc0 already fixed — returns `null` when there's no padding.

File: packages/super-editor/src/editors/v1/extensions/table-cell/helpers/parseCellMargins.js (hunk 1 above)
Comment (caio-pizzol) [reply packages/super-editor/src/editors/v1/extensions/table-cell/helpers/parseCellMargins.js]: @Branc0 the formula is right, just the name is confusing — we mentioned a shared helper in the inline review.

File: packages/super-editor/src/editors/v1/extensions/table-cell/table-cell.js (hunk 2 above)
Comment (caio-pizzol) [reply packages/super-editor/src/editors/v1/extensions/table-cell/table-cell.js]: makes sense, `middle` → `center` looks good.

File: packages/super-editor/src/editors/v1/extensions/table-row/helpers/parseRowHeight.js (hunk 3 above)
Comment (caio-pizzol) [reply packages/super-editor/src/editors/v1/extensions/table-row/helpers/parseRowHeight.js:81]: good approach, works correctly in testing.

File: packages/super-editor/src/editors/v1/extensions/shared/parseCellBorders.js (hunk 4 above)
Comment (caio-pizzol) [reply packages/super-editor/src/editors/v1/extensions/shared/parseCellBorders.js:34]: fine to keep them, no harm having extra styles for other paste sources.

File: packages/layout-engine/painters/dom/src/table/renderTableRow.ts (hunk 5 above)
Comment (caio-pizzol) [reply packages/layout-engine/painters/dom/src/table/renderTableRow.ts:91]: answered this in the inline review — GDocs always sets both sides so no risk here.

**Comment by caio-pizzol** [comment]
Hey @Branc0, this PR has had no activity for 3 days.

File: packages/super-editor/src/editors/v1/extensions/shared/parseCellMargins.js (hunk 8 above)
Comment (Branc0) [reply packages/super-editor/src/editors/v1/extensions/shared/parseCellMargins.js:1]: @caio-pizzol, we briefly spoke about this on Friday morning, but I forgot the outcome of our discussion 😅. 

`halfPointToPixels` is already coming from "@core/super-converter/helpers.js", do you want me to change its name, or keep it as it is?

File: packages/layout-engine/painters/dom/src/table/renderTableRow.ts (hunk 7 above)
Comment (Branc0) [reply packages/layout-engine/painters/dom/src/table/renderTableRow.ts:85]: Okay, I just added a comment there. Let me know if it's okay.

File: packages/super-editor/src/editors/v1/extensions/shared/parseCellBorders.js (hunk 6 above)
Comment (Branc0) [reply packages/super-editor/src/editors/v1/extensions/shared/parseCellBorders.js:1]: I added more unitary tests and created some integration tests

**Review by caio-pizzol — COMMENTED** [review]
@Branc0 tests and comment look good, resolved those two threads.

one thing: Google Docs uses `<th>` for header rows, but `table-header.js` only has rendering — it's missing the parsing side for the new attributes. so when you paste a table with a header row, the header cells lose their padding (and would lose background/borders/alignment too if they had any). `table-cell.js` has both sides, `table-header.js` needs the same. tested with [this doc](https://docs.google.com/document/d/18PnYvrj60RuJ_s-IyidzQs0BodfPu0-SCqv_4SIKCkg/edit?usp=sharing).

re: the `halfPointToPixels` name — keep it as is, it's fine.

File: packages/super-editor/src/editors/v1/extensions/shared/parseCellMargins.js (hunk 8 above)
Comment (caio-pizzol) [reply packages/super-editor/src/editors/v1/extensions/shared/parseCellMargins.js:1]: keep the name, it's fine. the duplication is minor — not worth extracting for this PR.

**Comment by Branc0** [comment]
Made some more changes:
- Moved the parsing functions  from `extensions/table-cell/helpers`  to `extensions/shared`
- Added the parsers for table-headers
- Added more integration tests + table-headers tests

**Review by caio-pizzol — APPROVED** [review]
@Branc0 all good from last round.

caught one bug — pasted background colors were stored with a `#` prefix, but the renderer adds its own `#`, so colors came out as `##ffff00` (broken). pushed a fix to your branch (`6b14e6d8c`) that strips the extra `#` at parse time, plus a round-trip test to prevent it from coming back.

approving.
